File: symbsat-cpp/monom.cpp
#include "monom.h"

bool Monom::isdivisible(const Monom& other) const {
    if (other.isOne()) {
        return true;
    } else if (this->isOne()) {
        return false;
    } else {
        return mVars == (mVars | other.mVars);
    }

}

bool Monom::isrelativelyprime(const Monom& other) const {
    if (mVars == other.mVars) {
        return true;
    } else if (this->isOne()) {
        return true;
    } else {
        auto lcm = mVars | other.mVars;
        return (lcm ^ mVars) == other.mVars;
    }
}

bool Monom::operator<(const Monom& b) const {
    // start with 1 because 0-bit indicates zero/one
    size_t msize = mVars.size();
    for (size_t i = 1; i < msize; ++i) {
        // if (mVars[i] < b.mVars[i])
            // return true;
        // else if (mVars[i] > b.mVars[i])
            // return false;
        if (mVars[i] ^ b.mVars[i]) return b.mVars[i];
    }
    return false;
}

Monom operator*(const Monom& a, const Monom& b) {
    if (a.isOne()) {
        return b;
    }
    if (b.isOne()) {
        return a;
    }
    if (a.isZero() || b.isZero()) {
        return Monom();
    }
    Monom tmp;
    tmp.mVars = a.mVars | b.mVars;
    return tmp;
}

Monom operator/(const Monom& a, const Monom& b) {
    if (b.isOne()) {
        return a;
    }
    if (a.isOne()) {
        return Monom();
    }
    if (a == b) {
        // return 1
        return Monom(0);
    }
    if (!a.isdivisible(b)) {
        // return 0
        return Monom();
    }
    Monom tmp;
    tmp.mVars = a.mVars ^ b.mVars;
    return tmp;
}

std::ostream& operator<<(std::ostream& out, const Monom &a) {
    if (a.isZero()) {
        out << "0";
    } else if (a.isOne()) {
        out << "1";
    } else {
        out << "[ ";
        // start from 1 because first bit is for 0/1
        for (size_t i=1; i<a.mVars.size(); ++i) {
            if (a.mVars[i]) {
                out << i-1 << " ";
            }
        }
        out << "]";
    }
    return out;
}

File: symbsat-py/monom.py
"""Boolean Monomials."""
import itertools
import operator
from functools import partialmethod

from order import Lex


class Monom:

    def __init__(self, size, bits=None, vars=None, order=Lex):
        self.size = size
        self.order = order
        if bits is not None:
            self.bits = tuple(bits)
        elif vars is not None:
            bits = [0] * size
            for var in vars:
                bits[var] = 1
            self.bits = tuple(bits)
        else:
            self.bits = tuple()

    @classmethod
    def zero(cls):
        super(cls).__init__()
        return cls(size=0, bits=tuple())

    @classmethod
    def one(cls):
        return cls(size=1, bits=(0,))

    def is_zero(self):
        return self.bits == ()

    def is_one(self):
        return self.bits == (0,)

    def __mul__(self, other):
        if self.is_one():
            return other
        if other.is_one():
            return self
        if self.is_zero():
            return Monom.zero()
        if other.is_zero():
            return Monom.zero()

        return Monom(
            size=self.size,
            bits=map(operator.or_, self.bits, other.bits),
            order=self.order
        )

    def __truediv__(self, other):
        if other.is_one():
            return self
        if self.is_one():
            return Monom.zero()
        if self == other:
            return Monom.one()
        if not self.isdivisible(other):
            return Monom.zero()

        return Monom(
            size=self.size,
            bits=map(operator.xor, self.bits, other.bits),
            order=self.order
        )

    def __hash__(self):
        return hash(self.bits)

    def __eq__(self, other):
        return self.bits == other.bits

    def __lt__(self, other):
        return self.order.lt(self, other)

    def __le__(self, other):
        return NotImplemented

    def __gt__(self, other):
        return not self.order.lt(self, other)

    def __ge__(self, other):
        return NotImplemented

    def __str__(self):
        if self.is_one():
            return "1"
        if self.is_zero():
            return "0"
        return str(self.vars)

    def pprint(self, alphabet, op=""):
        return op.join(l for v, l in zip(self, alphabet) if v == 1)

    def lcm(self, other):
        return self*other

    def isdivisible(self, other):
        if other.is_one():
            return True
        if self.is_one():
            return False
        return self == Monom(
            size=self.size,
            bits=map(operator.or_, self.bits, other.bits),
            order=self.order
        )

    def isrelativelyprime(self, other):
        if self == other:
            return True
        if self.is_one():
            return True
        lcm = self.lcm(other)
        return Monom(
            size=self.size,
            bits=map(operator.xor, lcm.bits, self.bits),
            order=self.order
        ) == other

    def prolong(self, i):
        """Prolongation of the monomial m*x."""
        assert not self[i]
        self[i] = 1

    @property
    def vars(self):
        """Return list of variables positions."""
        return [i for i, v in enumerate(self.bits) if v == 1]

    @property
    def degree(self):
        return sum(self.bits)


def make_monom_type(size):

    class _Monom(Monom):
        __init__ = partialmethod(Monom.__init__, size)

        @classmethod
        def zero(cls):
            return cls(bits=tuple())

        @classmethod
        def one(cls):
            return cls(bits=(0,))
    return _Monom

Monom4 = make_monom_type(4)
